File: java-source/othello/OthelloModel.java
import java.util.ArrayList;
import java.util.List;
public class OthelloModel {
    private final int size = 8;
    private final Square[][] board;
    private Player currentPlayer;

    public OthelloModel(){
        board = new Square[size][size];
        resetGame();
    }

    public void resetGame(){
        for (int i = 0; i < size; i++){
            for (int j = 0; j < size; j++){
                board[i][j] = Square.EMPTY;
            }
        }
        board[3][3] = Square.WHITE;
        board[3][4] = Square.BLACK;
        board[4][3] = Square.BLACK;
        board[4][4] = Square.WHITE;
        currentPlayer = Player.BLACK;
    }
    public int getSize(){
        return size;
    }
    public Square getSquare(int row, int column){
        return board[row][column];
    }

    public Player getCurrentPlayer() {
        return currentPlayer;
    }
    public void switchTurn(){
        currentPlayer = opponent(currentPlayer);
    }
    public Player opponent(Player player){
        if(player == Player.BLACK){
            return Player.WHITE;
        }
        return Player.BLACK;
    }
    public Square playersSquare(Player player){
        if (player == Player.BLACK){
            return Square.BLACK;
        }
        return Square.WHITE;
    }
    public  Square opponentSquareColor(Player player){
        if( player == Player.BLACK){
            return Square.WHITE;
        }
        return Square.BLACK;
    }

    public  boolean isValidMove(int row, int column, Player player){
        if (isInside(row, column) == false){
            return false;
        }
        if (board[row][column] != Square.EMPTY){
            return false;
        }
        int[][] directions = getDirections();

        for (int[] direction : directions){
            if (capturesInAnyDirection(row, column, direction[0], direction[1], player) == true){
                return true;
            }
        }
        return false;
    }

    private boolean isInside(int row, int column){
        if ((row >= 0 && row < size && column >= 0 && column < size) == true){
            return true;
        }
        return false;
    }
    private int[][] getDirections(){
        int[][] directions = new int[][]{ {-1,-1}, {-1, 0}, {-1, 1}, {0,-1}, {0,1}, {1,-1}, {1, 0}, {1,1}};
        return directions;
    }

    private boolean capturesInAnyDirection(int row, int column, int directionRow, int directionColumn, Player player){
        int r = row + directionRow;
        int c = column + directionColumn;
        Square opponent = opponentSquareColor(player);
        Square own = playersSquare(player);
        boolean foundOpponent = false;

        while(isInside(r,c) == true){
            if(board[r][c] == opponent){
                foundOpponent = true;
            } else if (board[r][c] == own) {
                return foundOpponent;
            } else{
                return false;
            }
            r += directionRow;
            c += directionColumn;
        }
        return false;
    }

    private void flipInDirection(int row, int column, int directionRow, int directionColumn, Player player){
        int r = row + directionRow;
        int c = column + directionColumn;
        Square opponent = opponentSquareColor(player);
        Square own = playersSquare(player);

        while(isInside(r,c) == true && board[r][c] == opponent) {
            board[r][c] = own;
            r += directionRow;
            c += directionColumn;
        }
    }
    public List<Move> getValidMoves(Player player){
        List<Move> moves = new ArrayList<>();
        for (int i = 0; i < size; i++){
            for (int j = 0; j < size; j++){
                if (isValidMove(i,j, player)){
                    moves.add(new Move(i,j));
                }
            }
        }
        return moves;
    }
    public boolean makeMove(int row, int column, Player player){
        if (isValidMove(row, column, player) == false){
            return false;
        }
        board[row][column] = playersSquare(player);
        int[][] directions = getDirections();
        for (int[] direction : directions){
            if (capturesInAnyDirection(row, column, direction[0], direction[1], player) == true){
                flipInDirection(row, column, direction[0], direction[1], player);
            }
        }
        currentPlayer = opponent(player);
        handleSkips();
        return true;
    }

    private void handleSkips(){
        if (getValidMoves(currentPlayer).isEmpty() && isGameOver() == false){
            currentPlayer = opponent(currentPlayer);
        }
    }
    public int countPieces(Player player){
        Square correctColor = playersSquare(player);
        int count = 0;
        for (int i = 0; i < size; i++){
            for (int j = 0; j < size; j++){
                if (board[i][j] == correctColor){
                    count++;
                }
            }
        }
        return count;
    }
    public boolean isGameOver(){
        if (getValidMoves(Player.BLACK).isEmpty() && getValidMoves(Player.WHITE).isEmpty()){
            return true;
        }
        return false;
    }
    public Player getWinner(){
        int black = countPieces(Player.BLACK);
        int white = countPieces(Player.WHITE);

        if (black > white){
            return Player.BLACK;
        } else if (white > black) {
            return Player.WHITE;
        } else {
            return null;
        }
    }

    public int checkWhoIsWinning(Player player){
        Player opponent = opponent(player);
        int score = countPieces(player) - countPieces(opponent);
        return score;
    }

    public OthelloModel copyBoard(){
        OthelloModel copy = new OthelloModel();
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                copy.board[i][j] = this.board[i][j];
            }
        }
        copy.currentPlayer = this.currentPlayer;
        return copy;
    }

    public enum Square{
        EMPTY, WHITE, BLACK
    }

    public enum Player{
        WHITE, BLACK
    }

    public static class Move {
        public final int row;
        public final int column;

        public Move(int row, int column) {
            this.row = row;
            this.column = column;
        }

        @Override
        public String toString(){
            return "(" + row + ", " + column + ")";
        }
    }

}

File: thearcadefinalfinal/games/othello/js/othello-model.js
/* ==========================================================================
   othello-model.js  —  port of OthelloModel.java
   Rules per https://www.ultraboardgames.com/othello/game-rules.php:
   black moves first, and a player with no legal move forfeits the turn.
   ========================================================================== */
(function () {
  "use strict";
  window.Oth = window.Oth || {};

  var Square = Object.freeze({ EMPTY: 0, WHITE: 1, BLACK: 2 });
  var Player = Object.freeze({ WHITE: "WHITE", BLACK: "BLACK" });

  var DIRECTIONS = [
    [-1, -1], [-1, 0], [-1, 1],
    [ 0, -1],          [ 0, 1],
    [ 1, -1], [ 1, 0], [ 1, 1]
  ];

  function Move(row, column) {
    this.row = row;
    this.column = column;
  }
  Move.prototype.toString = function () {
    return "(" + this.row + ", " + this.column + ")";
  };

  function OthelloModel() {
    this.size = 8;
    this.board = [];
    for (var i = 0; i < this.size; i++) this.board.push(new Int8Array(this.size));
    this.currentPlayer = Player.BLACK;
    this.resetGame();
  }

  OthelloModel.prototype.resetGame = function () {
    for (var i = 0; i < this.size; i++) {
      for (var j = 0; j < this.size; j++) this.board[i][j] = Square.EMPTY;
    }
    this.board[3][3] = Square.WHITE;
    this.board[3][4] = Square.BLACK;
    this.board[4][3] = Square.BLACK;
    this.board[4][4] = Square.WHITE;
    this.currentPlayer = Player.BLACK;   // black goes first in Othello
  };

  OthelloModel.prototype.getSize = function () { return this.size; };

  OthelloModel.prototype.getSquare = function (row, column) {
    return this.board[row][column];
  };

  OthelloModel.prototype.getCurrentPlayer = function () { return this.currentPlayer; };

  OthelloModel.prototype.switchTurn = function () {
    this.currentPlayer = this.opponent(this.currentPlayer);
  };

  OthelloModel.prototype.opponent = function (player) {
    return player === Player.BLACK ? Player.WHITE : Player.BLACK;
  };

  OthelloModel.prototype.playersSquare = function (player) {
    return player === Player.BLACK ? Square.BLACK : Square.WHITE;
  };

  OthelloModel.prototype.opponentSquareColor = function (player) {
    return player === Player.BLACK ? Square.WHITE : Square.BLACK;
  };

  OthelloModel.prototype.isInside = function (row, column) {
    return row >= 0 && row < this.size && column >= 0 && column < this.size;
  };

  OthelloModel.prototype.isValidMove = function (row, column, player) {
    if (!this.isInside(row, column)) return false;
    if (this.board[row][column] !== Square.EMPTY) return false;

    for (var d = 0; d < DIRECTIONS.length; d++) {
      if (this.capturesInAnyDirection(row, column, DIRECTIONS[d][0], DIRECTIONS[d][1], player)) {
        return true;
      }
    }
    return false;
  };

  /* Walk out from (row, column): we need at least one enemy disc, then one of
     ours, with no gap in between. */
  OthelloModel.prototype.capturesInAnyDirection = function (row, column, dRow, dCol, player) {
    var r = row + dRow;
    var c = column + dCol;
    var opponent = this.opponentSquareColor(player);
    var own = this.playersSquare(player);
    var foundOpponent = false;

    while (this.isInside(r, c)) {
      if (this.board[r][c] === opponent) {
        foundOpponent = true;
      } else if (this.board[r][c] === own) {
        return foundOpponent;
      } else {
        return false;
      }
      r += dRow;
      c += dCol;
    }
    return false;
  };

  OthelloModel.prototype.flipInDirection = function (row, column, dRow, dCol, player) {
    var r = row + dRow;
    var c = column + dCol;
    var opponent = this.opponentSquareColor(player);
    var own = this.playersSquare(player);

    while (this.isInside(r, c) && this.board[r][c] === opponent) {
      this.board[r][c] = own;
      r += dRow;
      c += dCol;
    }
  };

  OthelloModel.prototype.getValidMoves = function (player) {
    var moves = [];
    for (var i = 0; i < this.size; i++) {
      for (var j = 0; j < this.size; j++) {
        if (this.isValidMove(i, j, player)) moves.push(new Move(i, j));
      }
    }
    return moves;
  };

  OthelloModel.prototype.makeMove = function (row, column, player) {
    if (!this.isValidMove(row, column, player)) return false;

    this.board[row][column] = this.playersSquare(player);
    this.lastMove = { row: row, column: column };

    for (var d = 0; d < DIRECTIONS.length; d++) {
      if (this.capturesInAnyDirection(row, column, DIRECTIONS[d][0], DIRECTIONS[d][1], player)) {
        this.flipInDirection(row, column, DIRECTIONS[d][0], DIRECTIONS[d][1], player);
      }
    }
    this.currentPlayer = this.opponent(player);
    this.handleSkips();
    return true;
  };

  /* If the player to move has nothing legal, the turn bounces back. */
  OthelloModel.prototype.handleSkips = function () {
    this.skipped = false;
    if (this.getValidMoves(this.currentPlayer).length === 0 && !this.isGameOver()) {
      this.currentPlayer = this.opponent(this.currentPlayer);
      this.skipped = true;
    }
  };

  OthelloModel.prototype.countPieces = function (player) {
    var correctColor = this.playersSquare(player);
    var count = 0;
    for (var i = 0; i < this.size; i++) {
      for (var j = 0; j < this.size; j++) {
        if (this.board[i][j] === correctColor) count++;
      }
    }
    return count;
  };

  OthelloModel.prototype.isGameOver = function () {
    return this.getValidMoves(Player.BLACK).length === 0 &&
           this.getValidMoves(Player.WHITE).length === 0;
  };

  OthelloModel.prototype.getWinner = function () {
    var black = this.countPieces(Player.BLACK);
    var white = this.countPieces(Player.WHITE);
    if (black > white) return Player.BLACK;
    if (white > black) return Player.WHITE;
    return null;                       // tie
  };

  /* Simple material evaluation: my discs minus theirs. */
  OthelloModel.prototype.checkWhoIsWinning = function (player) {
    return this.countPieces(player) - this.countPieces(this.opponent(player));
  };

  /* The Java copyBoard() went through `new OthelloModel()`, which resets the
     whole board just to overwrite it a line later. Minimax calls this tens of
     thousands of times, so here it clones directly off the prototype instead —
     same resulting object, none of the setup cost. */
  OthelloModel.prototype.copyBoard = function () {
    var copy = Object.create(OthelloModel.prototype);
    copy.size = this.size;
    copy.board = new Array(this.size);
    for (var i = 0; i < this.size; i++) copy.board[i] = this.board[i].slice();
    copy.currentPlayer = this.currentPlayer;
    return copy;
  };

  window.Oth.Square = Square;
  window.Oth.Player = Player;
  window.Oth.Move = Move;
  window.Oth.OthelloModel = OthelloModel;
})();
